File: leetcode/42.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_area = 0

        while right > left:
            max_area = max(max_area, min(height[left], height[right]) * (right-left))

            if height[left] < height[right]:
                    left += 1
            elif height[right] < height[left]:
                    right -= 1
            else:
                while height[right] == height[left] and right > left:
                    left += 1

        return max_area

File: leetcode-new/42.cpp
class Solution
{
public:
    int trap(vector<int> &height)
    {
        int res = 0;
        int n = height.size();
        if (n == 1)
            return 0;

        int i = 0;
        int j = 0;
        int rt = 0;
        while (++j < n)
        {
            if (height[j] >= height[i])
            {
                res += rt;
                rt = 0;
                i = j;
            }
            else
                rt += height[i] - height[j];
        }

        int sp = i;
        i = n - 1;
        j = n - 1;
        rt = 0;
        while (--i >= sp)
        {
            if (height[i] >= height[j])
            {
                res += rt;
                rt = 0;
                j = i;
            }
            else
                rt += height[j] - height[i];
        }
        return res;
    }
};
